src/cmap/conn/mod.rs

```diff
@@ -340,6 +377,21 @@ impl Drop for Connection {
     }
 }
 
+#[derive(Clone, Debug)]
```

Comment:
Blergh, well spotted.

For the code as currently written, I _think_ this can only happen via `timeout` or some similar future cancellation; if the `GetMoreProvider` future is polled to completion, it'll return the pinned connection and it'll be available for the cursor's `drop` invocation of `kill_cursors`.  Concurrent _execution_ of `getMore` with a cursor's `drop` can't happen, I think, because a normal `Cursor` is its own stream and a `SessionCursorStream` can't outlive the `SessionCursor`.

Given that we've already documented that cancelling futures returned by the driver isn't supported, I think getting an internal error if you do isn't a terrible outcome.

For the more general issue, yeah, it's definitely unfortunately easy to accidentally clone and overuse a handle.  I updated the `PinnedConnection[Handle]` types to not implement `Clone` and instead have a `replicate()` method that does the same thing, with a comment on what it's for; that should at least prevent accidentally cloning via including in other structs or the like.

Reply:
I actually think we may have races here even without cancellation, e.g.
```rust
cursor.next().await; // after this poll completes, connection is sent back over a channel via `store_pinned`
cursor.next().await; // `take_connection` races with the return of the connection
```
I think we may need to make returning a pinned connection an acknowledged operation, so that we can be sure the next time the connection is used it will be available. Alternatively, we can make `take_connection` block until its available or perished.

>For the more general issue, yeah, it's definitely unfortunately easy to accidentally clone and overuse a handle. I updated the PinnedConnection[Handle] types to not implement Clone and instead have a replicate() method that does the same thing, with a comment on what it's for; that should at least prevent accidentally cloning via including in other structs or the like.

This seems like a nice improvement :+1:

Reply:
I realized that pinning can use a channel directly between the handle and the connection itself, removing the pool worker from the path and simplifying things all around - this should fix that race since `take_connection` now blocks.  The downside of this is that calling `take_connection` in the same task that's holding the pinned `Connection` will deadlock rather than causing an internal error; I think this is okay because that'll show up immediately in spec tests.